File: frontend/src/utils/passwordRules.js
export const PASSWORD_RULES = [
  'Minimum 8 characters',
  'At least 1 uppercase letter',
  'At least 1 number',
  'At least 1 special symbol',
];

export function validatePassword(password) {
  const value = password || '';
  const checks = {
    minLength: value.length >= 8,
    uppercase: /[A-Z]/.test(value),
    number: /\d/.test(value),
    special: /[^A-Za-z0-9]/.test(value),
  };

  return {
    valid: Object.values(checks).every(Boolean),
    checks,
    message: 'Password must be at least 8 characters and include an uppercase letter, a number, and a special symbol.',
  };
}

File: backend/backend/src/main/java/com/stazy/backend/common/util/PasswordRules.java
package com.stazy.backend.common.util;

import com.stazy.backend.common.exception.BadRequestException;
import java.util.regex.Pattern;

public final class PasswordRules {

    private static final Pattern UPPERCASE = Pattern.compile("[A-Z]");
    private static final Pattern DIGIT = Pattern.compile("\\d");
    private static final Pattern SPECIAL = Pattern.compile("[^A-Za-z0-9]");

    private PasswordRules() {
    }

    public static void validate(String password) {
        if (password == null || password.length() < 8) {
            throw new BadRequestException("Password must be at least 8 characters long.");
        }
        if (!UPPERCASE.matcher(password).find()) {
            throw new BadRequestException("Password must contain at least 1 uppercase letter.");
        }
        if (!DIGIT.matcher(password).find()) {
            throw new BadRequestException("Password must contain at least 1 number.");
        }
        if (!SPECIAL.matcher(password).find()) {
            throw new BadRequestException("Password must contain at least 1 special symbol.");
        }
    }
}
